### DISCUSSION TITLE
The bare minimum to start with

### DISCUSSION THREAD

**Discussion by DZsolt01** [autobase-tech/autobase#651] [General]
Hello,
What will be the bare minimum number of nodes to start with?
Nodes in this case means VPS.
The basic idea would be to make a budget start.
Use case: (Hetzner VPS provider)
I will attach an image which represent the current idea..
I am a web developer and want to manage my clients application and I come up with a small infrastructure which I think it should be enough for my 1-2 clients. (Mostly the applications are Node.js based, Next.js-frontend, Nest.js - backend)
My question would be it is a possibility for this arhitecture to work? 
Later on I would like to add more to every part of it. Just for the start I want to keep it small.
![image](https://github.com/vitabaks/postgresql_cluster/assets/79136090/743e7221-dcaa-4687-828a-5303c0d80501)

**Comment by vitabaks** [comment]
Hello @DZsolt01 

A minimum of 3 servers is required to ensure high availability. This requirement primarily applies to DCS (etcd), why this is so can be viewed [here](http://thesecretlivesofdata.com/raft/) 

The database cluster itself can consist of at least 2 nodes. 

In your case, you could (in order to save costs - to start) deploy etcd cluster on database servers (2 nodes) + 1 small VM for the third node of the DCS cluster. or even install 1 etcd on the application server. But I recommend thinking about dedicated VMs for the etcd cluster for production.

**Comment by vitabaks** [comment]
also, your diagram shows the HAProxy node, but since there is only one, this is a bottleneck. Consider installing it on database servers, or (better) use a Hetzner load balancer. 

I'm currently working on a cloud integration PR (https://github.com/vitabaks/postgresql_cluster/pull/464) and here's how you can deploy a cluster of 3 nodes + Hetzner load balancer with one command:

```
export HCLOUD_API_TOKEN=*********

ansible-playbook deploy_pgcluster.yml \
  --user=root --private-key=$HOME/.ssh/id_rsa --extra-vars \
    "cloud_provider=hetzner \
     server_count=3 \
     server_type=CCX13 \
     server_image=ubuntu-22.04 \
     server_location=ash \
     volume_size=100 \
     ssh_public_keys=\"$(cat $HOME/.ssh/id_rsa.pub)\"
     pgbouncer_install=false \
     postgresql_version=16 \
     patroni_cluster_name=postgres-cluster \
     cloud_load_balancer=true \
     database_public_access=false"
```

Note: Install PgBouncer pooler (pgbouncer_install=true) if you need it for your project, or start without it and you can add it later.

Example: https://github.com/vitabaks/postgresql_cluster/pull/464#issuecomment-1999688833

**Reply by DZsolt01** [reply]
Thanks for quick response. 
So it would be better to have 3 vps nodes for the databases, every one of them have etcd as well (at start later on move it to it's own cluster). And have a Hetzner load balancer on top of the databases? If I understood correctly like this:
![image](https://github.com/vitabaks/postgresql_cluster/assets/79136090/5f0a4e7b-023b-4e72-9384-7bd4976a9d94)
Sorry If I understood something wrong, I'm a beginner in this topic, I would like to learn :)

**Reply by vitabaks** [reply]
All right. Just run the command above (use the code from the 'cloud' branch) and you will be happy :)

**Reply by DZsolt01** [reply]
Hello @vitabaks , I started to run the config right now and I run into a problem with deployment ansible script.
![image](https://github.com/vitabaks/postgresql_cluster/assets/79136090/d7b0efa6-2d68-47da-91e3-df3aeeb95b9c)

**Reply by vitabaks** [reply]
please attach full ansible log

**Reply by DZsolt01** [reply]
Sorry, I got that one solved

**Reply by DZsolt01** [reply]
This is the new error:
[error.txt](https://github.com/vitabaks/postgresql_cluster/files/15224429/error.txt)

**Reply by DZsolt01** [reply]
@vitabaks do you have any idea?

**Reply by vitabaks** [reply]
The log is difficult to read because it displays all the tasks, even those that are skipped. It looks like you are using a different ansible configuration. I recommend the [display_skipped_hosts = False](https://github.com/vitabaks/postgresql_cluster/blob/master/ansible.cfg#L8) parameter so that the log is cleaner.

Try to delete the cluster (run the same command but add the `state=absent` variable) and re-deploy it.

**Reply by vitabaks** [reply]
You can specify your SSH key that was previously uploaded to the Cloud. 

Try using the `ssh_key_name=<key_name>` variable instead of ssh_public_keys. Then there will be no attempt to create a new SSH key for deployment, but the one that you specify will be used.

**Reply by DZsolt01** [reply]
Seems it's worked out
![image](https://github.com/vitabaks/postgresql_cluster/assets/79136090/1bcd563a-3604-430f-b97b-6528d86aa0b1)

**Reply by DZsolt01** [reply]
@vitabaks It's working. Thanks a lot! Any suggestions for the future regarding to the infrastructure? Any good practices should I follow?

**Reply by vitabaks** [reply]
>  Any suggestions for the future regarding to the infrastructure? 

To a greater extent, it depends on your service, and its growth rate. \
At a minimum, if the load increases, you may need to change the server type to a more powerful one.  

Make sure in advance that your application skillfully sends reading requests to the address of the read-only load balancer, so you horizontally distribute the load across the cluster.

> Any good practices should I follow?

Study carefully the possibilities of this automation, see [README](https://github.com/vitabaks/postgresql_cluster/blob/master/README.md) (including links).

- Play around with the cluster, and try to break it.
- Set up database backups
- Set up a database monitoring (you can set [Netdata](https://www.netdata.cloud) by specifying the `netdata_install=true`  variable). I also recommend https://gitlab.com/postgres-ai/pgwatch2 
- study the documentation of each component that makes up the cluster

Or consider participating in a sponsorship program via [GitHub](https://github.com/sponsors/vitabaks) or [Patreon](https://patreon.com/vitabaks) that also includes personal support including live chat with me on a dedicated Slack channel.

**Comment by vitabaks** [comment]
> I am a web developer

hmm, please take a look at this issue https://github.com/vitabaks/postgresql_cluster/issues/589

You may wish to participate in this project.

**Reply by DZsolt01** [reply]
Currently my time is too limited. In the future I would like to participate. (In like 2 months)

**Reply by vitabaks** [reply]
Ok, I will be glad if you join.
